### ISSUE TITLE
EventWriterT change 

### ISSUE THREAD

**Issue by Saulzar** [reflex-frp/reflex#233]
EventWriterT was recently changed so that it's `tellEvent`is now prompt with regards to runWithReplace, that is that if a `tellEvent e` co-occurs when it is switched out by the same event `e` then the tellEvent does not occur.

https://github.com/reflex-frp/reflex/commit/ca50bb4cf1e85e1c980b3ddad0e1b0c483542349#diff-ca34969f303705bc74cc027f0eedb19f

I would argue that in a lot of cases runWithReplace it would be desirable to have `tellEvent` be non promptly switched, does this require two kinds of EventWriterT one for prompt and one for non prompt switching?

**Comment by Saulzar** [comment]
After more investigating I'm not even sure quite what the behaviour of the new changes is. I have a lot of uses of this pattern where I have some code which looks like this:

```
state = Workflow $ do
    ....
    tellEvent foo
    ...
    return (x, baseState <$ foo)

```

Most of them seem to continue working (contrary to the description of the new behaviour). One or two of them have stopped working and I'm not really sure why. Will see if I can find an isolated test case.

**Comment by Saulzar** [comment]
FWIW. In one case I have this code:

```
performEvent (liftIO . print <$> e1)
tellEvent e1

```

This seems to only occur when `e1` is the result of a switch from some other event `e2` and `e2` is causing this widget to be replaced. So perhaps EventWriterT has the correct behavior here.
The performEvent prints, but tellEvent does not fire - at the least this is inconsistent?

**Comment by JBetz** [comment]
Possibly related. since upgrading to `reflex-0.5`, a lot of my widgets that use `tellEvent` are broken. The event only gets sent up the first time it's triggered. 

For example:

```haskell
dyn_ $ ffor stateD $ \(State _) -> do
... 
tellEvent $ pure <$> select ZoomIn menuE
tellEvent $ pure <$> select ZoomOut menuE
...
```

I can only "tell" one of those events, and it can only be "told" once.

Is this related to the change you mentioned?

**Comment by ryantrinkle** [comment]
@JBetz I believe @Saulzar's case would apply if `stateD` is changing at the same time that `menuE` is firing.  However, it doesn't explain why you'd be seeing only a single `tellEvent` taking effect.  Would you mind posting a bit more context? I'd like to see if I can reproduce the issue and get it fixed ASAP.

@Saulzar Yes, that semantic change was intentional, as I believe the previous behavior didn't make much sense.  However, I may have made the wrong call to release it without some kind of deprecation cycle.  Sorry for making you debug through this!  I will think about what to do about `RequesterT` - thanks for bringing it up.

**Comment by JBetz** [comment]
@ryantrinkle I think it's an issue with `tellEvent` being used inside of a dynamic produced by `listWithKey`. Here's a MVE:

```haskell
testW :: MonadWidget t m => m ()
testW = mdo
  let update [()] cur = cur + 1
      update _ cur    = cur
  counterD <- foldDyn update 0 counterE
  (_, counterE) <- runEventWriterT $
    listWithKey (Map.singleton 0 <$> counterD) $ \_ counterD -> do
      dynText $ pack . show <$> counterD
      incE <- button "+"
      tellEvent $ pure <$> incE
  pure ()
```

**Comment by saulzar** [comment]
Much like the different variations in switch/switchPromptly/switchPromptlyOnly  (and much as it is painful to have umpteen variations of the same thing), is it reasonable to have both prompt/non prompt switching EventWriterT? 

Personally I prefer non prompt switching to be a more intuitive default, it also has an advantage of being simpler (and more performant).

**Comment by ryantrinkle** [comment]
Yes, but for what it's worth, switchPromptlyOnly should be the only version that exists; `switchPromptly` is basically buggy, but I was worried to change its behavior without notice since it seemed impossible for people to debug.  We should do a (long) deprecation cycle to rename switchPromptlyOnly to switchPromptly and drop the current version of switchPromptly.

**Comment by saulzar** [comment]
That _is_ strange, and I agree switchPromptOnly should be the 'normal' version. If the other version did exist it should only be with Monoid or These to highlight the overlap!

**Comment by Ericson2314** [comment]
I'm curious what affect #371will have on this

**Comment by JBetz** [comment]
@Ericson2314 There's a [commented out test case](https://github.com/reflex-frp/reflex/blob/develop/test/RequesterT.hs#L57) in the RequesterT test suite with a TODO about re-enabling it once this issue has been resolved. The test passes on the #371 branch, so it _looks_ like this will be fixed by the refactor.
